File: Packages/StewardMDWatchCore/Sources/StewardMDWatchCore/Engines/CompressionAnalyzer.swift
import Foundation

/// One acceleration sample. `value` is a 1-D compression signal — the signed
/// vertical component of user acceleration (userAcceleration projected onto gravity),
/// which oscillates once per compression (unlike |acceleration|, which peaks twice
/// per stroke and double-counts).
public struct CompressionSample: Sendable, Equatable {
    public let t: TimeInterval
    public let value: Double
    public init(t: TimeInterval, value: Double) { self.t = t; self.value = value }
}

/// Measured compression state (design §3.1). No quality/depth — counts + rate + pause only.
public struct CompressionState: Sendable, Equatable {
    public var count = 0
    public var instantaneousRateCPM = 0
    public var averageRateCPM = 0
    public var paused = false
    public var pauseSeconds: TimeInterval = 0
    public init() {}
}

/// What a single `ingest` produced, so the model can emit `CodeEvent`s with ids.
public struct AnalyzerTick: Sendable, Equatable {
    public var compressionCounted = false
    public var pauseStarted = false
    public var resumed = false
    public init() {}
}

/// Counts chest compressions + estimated rate from a 1-D acceleration signal
/// (design §3.1). Pure and deterministic — no CoreMotion, no clock; advanced only by
/// `ingest`.
///
/// Accuracy design (fixes one-compression-counted-twice):
///  • **Hysteresis / Schmitt trigger** — a compression is counted only when the
///    smoothed signal rises above an upper threshold *after* having fallen below a
///    lower threshold. A secondary bump within the same stroke never dips below the
///    low threshold, so it can't add a second count.
///  • **Adaptive thresholds** — high/low track a slow baseline ± a fraction of the
///    running amplitude, so it self-scales to how hard/deep the compressions are.
///  • **Low-pass smoothing** removes jitter that would otherwise fragment a peak.
///  • **Refractory backstop** caps the max plausible rate.
///  • **Repetition gate** withholds counting until several rhythmic compressions are
///    seen (rejects a single accidental movement).
///  • **Pause detection** — no compression for `pauseTimeout` → paused; auto-resume.
public struct CompressionAnalyzer: Sendable {
    public var refractory: TimeInterval
    public var pauseTimeout: TimeInterval
    public var repetitionGate: Int
    public var minAmplitude: Double     // g — floor so flat noise never crosses the gate
    public var hysteresis: Double       // fraction of amplitude for the high/low band
    public var smoothing: Double        // EMA alpha for the input (0…1; higher = less smoothing)

    public private(set) var state = CompressionState()

    private var smoothed: Double?
    private var baseline = 0.0
    private var dev = 0.2
    private var armed = false               // has dipped below `low` since the last count
    private var lastPeakT: TimeInterval?
    private var counting = false            // repetition gate satisfied
    private var provisional = 0
    private var intervals: [TimeInterval] = []

    public init(refractory: TimeInterval = 0.28, pauseTimeout: TimeInterval = 3.0,
                repetitionGate: Int = 3, minAmplitude: Double = 0.15,
                hysteresis: Double = 0.5, smoothing: Double = 0.35) {
        self.refractory = refractory; self.pauseTimeout = pauseTimeout
        self.repetitionGate = repetitionGate; self.minAmplitude = minAmplitude
        self.hysteresis = hysteresis; self.smoothing = smoothing
    }

    public mutating func ingest(_ s: CompressionSample) -> AnalyzerTick {
        var tick = AnalyzerTick()

        // Low-pass smooth, then track a slow baseline + running amplitude.
        let x = smoothed.map { $0 + smoothing * (s.value - $0) } ?? s.value
        smoothed = x
        baseline += 0.02 * (x - baseline)
        dev += 0.05 * (abs(x - baseline) - dev)
        let amp = max(minAmplitude, dev)
        let high = baseline + hysteresis * amp
        let low = baseline - hysteresis * amp

        // Pause: no counted compression within the timeout.
        if let lp = lastPeakT {
            let gap = s.t - lp
            if !state.paused, gap >= pauseTimeout {
                state.paused = true; tick.pauseStarted = true; state.instantaneousRateCPM = 0
            }
            if state.paused { state.pauseSeconds = gap }
        }

        // Schmitt trigger: arm on a low excursion, count on the next high crossing.
        if x < low { armed = true }
        if x > high, armed {
            armed = false
            let interval = lastPeakT.map { s.t - $0 }
            if interval == nil || interval! >= refractory {
                register(at: s.t, interval: interval, tick: &tick)
            }
        }
        return tick
    }

    private mutating func register(at t: TimeInterval, interval: TimeInterval?, tick: inout AnalyzerTick) {
        let rhythmic = interval.map { $0 >= 0.30 && $0 <= 1.5 } ?? false   // 40–200 cpm

        if state.paused {
            state.paused = false; state.pauseSeconds = 0; tick.resumed = true
            intervals.removeAll()
        }

        if !counting {
            if rhythmic || provisional == 0 { provisional += 1 } else { provisional = 1 }
            if provisional >= repetitionGate {
                counting = true
                state.count = provisional        // backfill the gate compressions
                tick.compressionCounted = true
            }
        } else {
            state.count += 1
            tick.compressionCounted = true
        }

        if let iv = interval, rhythmic {
            state.instantaneousRateCPM = Int((60.0 / iv).rounded())
            intervals.append(iv)
            if intervals.count > 8 { intervals.removeFirst(intervals.count - 8) }
            let mean = intervals.reduce(0, +) / Double(intervals.count)
            state.averageRateCPM = mean > 0 ? Int((60.0 / mean).rounded()) : 0
        }
        lastPeakT = t
    }
}

File: android/wear/src/main/kotlin/in/stewardmd/wear/codeblue/CompressionAnalyzer.kt
package `in`.stewardmd.wear.codeblue

/**
 * Chest-compression detector. Port of WatchCore CompressionAnalyzer. Consumes a 1-D (gravity-removed)
 * acceleration signal one sample at a time and reports COUNT + estimated RATE + PAUSE only.
 *
 * SAFETY (design §0): NO depth, NO quality/adequacy. There is deliberately no depth field or method on
 * this class — rate is the only measured signal surfaced.
 *
 * Pipeline: 1-pole low-pass (de-jitter) → rising→falling peak per compression, gated by an amplitude
 * threshold → refractory backstop caps the max plausible rate. Rate = 60 / median(recent intervals).
 *
 * [calibration] scales the amplitude gate: a real accelerometer's axis/scale varies per device, so this
 * stays a tunable knob (tune against a metronome on real hardware — Task 13), never hard-coded.
 */
class CompressionAnalyzer(
    private val sampleRateHz: Double = 50.0,
    private val calibration: Double = 1.0,
    private val maxRateCpm: Int = 150,      // refractory backstop (compressions can't plausibly exceed this)
    private val pauseSeconds: Double = 2.0,
    private val smoothing: Double = 0.3,    // 1-pole LPF factor
    private val baseThresholdG: Double = 0.05,
) {
    private var lpf = 0.0
    private var prevLpf = 0.0
    private var rising = false
    private var t = 0.0
    private var lastPeakT = Double.NEGATIVE_INFINITY
    private val intervals = ArrayDeque<Double>()

    var count = 0
        private set

    private val minInterval get() = 60.0 / maxRateCpm
    private val threshold get() = baseThresholdG * calibration

    /** Feed one acceleration sample (gravity-removed, e.g. Android TYPE_LINEAR_ACCELERATION magnitude
     *  along the compression axis). */
    fun add(sample: Double) {
        t += 1.0 / sampleRateHz
        lpf += smoothing * (sample - lpf)
        val slope = lpf - prevLpf
        val isPeak = slope < 0 && rising && lpf > threshold && (t - lastPeakT) >= minInterval
        if (isPeak) {
            if (lastPeakT.isFinite()) {
                intervals.addLast(t - lastPeakT)
                while (intervals.size > 6) intervals.removeFirst()
            }
            lastPeakT = t
            count++
        }
        rising = slope > 0
        prevLpf = lpf
    }

    /** Estimated compressions per minute (0 until enough peaks are seen). */
    val rateCpm: Int
        get() {
            if (intervals.isEmpty()) return 0
            val sorted = intervals.sorted()
            val median = sorted[sorted.size / 2]
            return if (median > 0) (60.0 / median).toInt() else 0
        }

    /** True when compressions have stopped (no peak within [pauseSeconds]). */
    val paused: Boolean
        get() = lastPeakT.isFinite() && (t - lastPeakT) > pauseSeconds
}
